### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by bohdan-mykhailenko**
[DEMO LINK](https://bohdan-mykhailenko.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

**Review by serentinos — APPROVED** [review]
Good job!

**Review by silchyn — CHANGES_REQUESTED** [review]
Should todos be completed initially?

**Comment by bohdan-mykhailenko** [comment]
> Should todos be completed initially?

I just set it for testing 'Clear completed' button

**Comment by bohdan-mykhailenko** [comment]
> Should todos be completed initially?

File: src/components/TodoFilter/TodoFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            selected: filterStatus === FilterStatus.completed,
+          },
+        )}
+        onClick={(event) => handleFilterClick(event, FilterStatus.completed)}
+      >
+        Completed
+      </a>
+    </nav>
+  );
+});
```
[the remark this answers is not in the document]
Comment (bohdan-mykhailenko) [reply src/components/TodoFilter/TodoFilter.tsx:61]: Good point
Here we go

const filterStatuses = Object
    .keys(FilterStatus) as (keyof typeof FilterStatus)[];

  return (
    <nav className="filter">
      {filterStatuses.map(status => (
        <Fragment key={status}>
          <a
            href={status !== 'all' ? `#/${status}` : '#/'}
            className={cn('filter__link', {
              selected: status === filterStatus,
            })}
            onClick={event => handleFilterClick(event, FilterStatus[status])}
          >
            {status}
          </a>
        </Fragment>
      ))}
    </nav>
  );
});

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,169 @@
...
+import { USER_ID } from './consts/consts';
+import './styles/todoapp.scss';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filterStatus, setFilterStatus] = useState(FilterStatus.all);
```
Comment (silchyn) [inline src/App.tsx:18]:
```suggestion
  const [statusFilter, setStatusFilter] = useState(FilterStatus.all);
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,169 @@
...
+  const [error, setError] = useState<string | null>(null);
+  const [isFetching, setIsFetching] = useState(false);
```
Comment (silchyn) [inline src/App.tsx:20]:
```suggestion
  const [isLoading, setIsLoading] = useState(false);
```

Comment (silchyn) [inline src/App.tsx:20]: What is fetching?

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,169 @@
...
+  const [title, setTitle] = useState('');
+  const [loadingTodoIds, setLoadingTodoIds] = useState([0]);
```
Comment (silchyn) [inline src/App.tsx:22]: Where this `0` comes from?

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,169 @@
...
+
+  useEffect(() => {
+    getTodos()
+      .then(todosFromServer => setTodos(todosFromServer))
```
Comment (silchyn) [inline src/App.tsx:26]:
```suggestion
      .then(setTodos)
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,169 @@
...
+      timerId = setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => clearTimeout(timerId);
+  }, [error]);
+
+  const todosCount = todos.length;
+  const isTodosVisible = todosCount > 0;
```
Comment (silchyn) [inline src/App.tsx:43]:
```suggestion
  const areTodosVisible = todosCount > 0;
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,24 +1,169 @@
...
+
+  const visibleTodos = useMemo(() => {
+    return getFilteredTodos(todos, filterStatus);
+  }, [todos, filterStatus]);
```
Comment (silchyn) [inline src/App.tsx:47]:
```suggestion
  const visibleTodos = useMemo(() => (
    getFilteredTodos(todos, filterStatus);
  ), [todos, filterStatus]);
```

File: src/App.tsx [hunk 8]
```diff
@@ -1,24 +1,169 @@
...
+
+  const completedTodos = useMemo(() => {
+    return todos.filter((todo) => todo.completed);
+  }, [todos, filterStatus]);
+
+  const isSomeActiveTodos = todosCount - completedTodos.length > 0;
```
Comment (silchyn) [inline src/App.tsx:53]:
```suggestion
  const hasActiveTodos = todosCount - completedTodos.length > 0;
```

Comment (silchyn) [inline src/App.tsx:53]: What if new status occurs?

**Review by silchyn — CHANGES_REQUESTED** [review]
Split logic into components

**Review by silchyn — APPROVED** [review]
The next part of this task will be reviewed

**Pull request closed without merging**